core-v1/src/main/scala/engine/SimpleCachingEngine.scala

```diff
@@ -0,0 +1,283 @@
+      val cb = ca.map(mapFn)
+      debugging(expr).invokeAndReturn(state(ca, cached(cb)))
+    }
+
+    override def visitNot[I, O : Negation : OP](expr: Expr.Not[I, O, OP]): I => CachedResult[O] =
+      memoize(expr, _) { i =>
+        val output = expr.innerExpr.visit(this)(i)
+        val negatedOutput = Negation[O].negation(output.value)
+        debugging(expr).invokeAndReturn(state((i, output.value), cached(negatedOutput, output.cacheState)))
+      }
+
+    override def visitOr[I](expr: Expr.Or[I, OP])(implicit evO: OP[Boolean]): I => CachedResult[Boolean] =
+      memoize(expr, _) { i =>
+        val (lo, ro, o) = visitWithUpdatedCache2(i, expr.leftExpr, expr.rightExpr) { (lo, ro) =>
+          lo || ro
+        }
+        debugging(expr).invokeAndReturn(state((i, lo, ro), o))
+      }
+
+    override def visitSelect[I, O : OP](expr: Expr.Select[I, O, OP]): I => CachedResult[O] = { i =>
+      // NOTE: This result is not memoized as select is simple to execute
+      val o = expr.lens.get(i)
+      debugging(expr).invokeAndReturn(state((i, expr.lens), cached(o)))
+    }
+
+    override def visitValuesOfType[T, O](
+      expr: Expr.ValuesOfType[T, O, OP],
+    )(implicit
+      opTs: OP[Seq[O]],
+    ): Any => CachedResult[Seq[O]] = { i =>
+      // NOTE: This result is not memoized as valuesOfType is simple to execute and does not depend on the input
+      val matchingFacts = factTable.getSortedSeq(expr.factTypeSet)
+      val o = matchingFacts.map(expr.transform)
+      debugging(expr).invokeAndReturn(state(i, cached(o)))
+    }
+
+    override def visitWithinWindow[I, V : OP, F[+_]](
+      expr: Expr.WithinWindow[I, V, F, OP],
+    )(implicit
+      comparison: WindowComparable[F, OP],
+      opB: OP[F[Boolean]],
+    ): I => CachedResult[F[Boolean]] = memoize(expr, _) { i =>
+      val (value, window, o) = visitWithUpdatedCache2(i, expr.valueExpr, expr.windowExpr) { (value, window) =>
+        comparison.withinWindow(value, window)
+      }
+      debugging(expr).invokeAndReturn(state((i, value, window), o))
+    }
+  }
+
+  def evalMultiple[OP[_]](
+    expressions: Seq[Expr[Any, Any, OP]],
+    factTable: FactTable = FactTable.empty,
+    cache: ResultCache = Map.empty,
+  ): IndexedSeq[Any] = {
+    val (results, cacheState) = expressions.foldLeft((Vector.empty[Any], cache)) {
+      case ((results, cache), next) =>
+        val result = next.visit(new Visitor[OP](factTable, cache))(())
+        (results :+ result.value, result.cacheState)
+    }
+    results
+  }
+
+  def debugMultiple[OP[_]](
+    expressions: Seq[Expr[Any, Any, OP]],
+    factTable: FactTable = FactTable.empty,
+    cache: ResultCache = Map.empty,
+  ): (IndexedSeq[CachedResult[Any]], ResultCache) = {
+    expressions.foldLeft((Vector.empty[CachedResult[Any]], cache)) {
+      case ((results, cache), next) =>
+        val result = next.visit(new Visitor[OP](factTable, cache))(())
+        (results :+ result, result.cacheState)
+    }
+  }
```

Comment:
I need to fold this back into the `RunExprDsl` somehow.

I should also support HList results using `NonEmptyExprHList` in a future PR. Maybe I should add a TODO as a reminder.